### PR TITLE
Additional image support: Concat and CompList and Expr

### PR DESCRIPTION
**Pull request by zarda**
From #388, add image support of CASA Concatenated image, CASA Componentlist image, and CASA Expression.

[Working test link](http://carta.asiaa.sinica.edu.tw/frontend/dev/?socketUrl=wss://carta.asiaa.sinica.edu.tw/socket5013)

Check list:

- Image type on file list
- File information
- Image readable

Testing image: set_CASAImage_flavors/

### CODE REVIEW

File: ImageData/CompListLoader.h [hunk 1]
```diff
@@ -0,0 +1,59 @@
...
+    std::string _filename;
+    std::unique_ptr<casacore::ImageInterface<casacore::Float> > _image;
+};
+
+CompListLoader::CompListLoader(const std::string& filename) : FileLoader(filename) {}
+
+void CompListLoader::OpenFile(const std::string& /*hdu*/) {
+    if (!_image) {
+        _image.reset(new casa::ComponentListImage(_filename));
+        _num_dims = _image->shape().size();
```
Comment (pford) [inline ImageData/CompListLoader.h:34]: Should check that _image is not nullptr before using it.

File: ImageData/ConcatLoader.h [hunk 2]
```diff
@@ -0,0 +1,60 @@
...
+    std::unique_ptr<casacore::ImageConcat<float>> _image;
+};
+
+ConcatLoader::ConcatLoader(const std::string& filename) : FileLoader(filename) {}
+
+void ConcatLoader::OpenFile(const std::string& /*hdu*/) {
+    if (!_image) {
+        casacore::JsonKVMap _jmap = casacore::JsonParser::parseFile(this->GetFileName() + "/imageconcat.json");
+        _image.reset(new casacore::ImageConcat<float>(_jmap, this->GetFileName()));
+        _num_dims = _image->shape().size();
```
Comment (pford) [inline ImageData/ConcatLoader.h:35]: Check that _image is not nullptr

File: ImageData/ExprLoader.h [hunk 3]
```diff
@@ -0,0 +1,65 @@
...
+ExprLoader::ExprLoader(const std::string& filename) : FileLoader(filename) {}
+
+void ExprLoader::OpenFile(const std::string& /*hdu*/) {
+    if (!_image) {
+        casacore::JsonKVMap _jmap = casacore::JsonParser::parseFile(_filename + "/imageexpr.json");
+        casacore::String _expr = _jmap.get("ImageExpr").getString();
+        casacore::PtrBlock<const casacore::ImageRegion*> _regions;
+        casacore::LatticeExprNode _node = casacore::ImageExprParse::command(_expr, casacore::Block<casacore::LatticeExprNode>(), _regions);
+        _image.reset(new casacore::ImageExpr<float>(_node, _expr, _filename, _jmap));
+        _num_dims = _image->shape().size();
```
Comment (pford) [inline ImageData/ExprLoader.h:40]: Check _image nullptr

**Review by pford — CHANGES_REQUESTED** [review]

**Comment by zarda** [comment]
Suggest to load the parent Loader::OpenFile(){ return default... } while child ___Loader::OpenFile() return nullptr, and to check nullptr before open file as well.

**Review by pford — APPROVED** [review]

**Pull request merged**